src/coreclr/jit/emitarm64.cpp

```diff
@@ -957,6 +957,208 @@ void emitter::emitInsSanityCheck(instrDesc* id)
             break;
     }
 }
+
+/*****************************************************************************
+ *
+ *  Sanity check two instruction are valid when placed next to each other
+ */
+
+void emitter::emitInsPairSanityCheck(instrDesc* firstId, instrDesc* secondId)
+{
+    if (firstId == nullptr || secondId == nullptr)
+    {
+        return;
+    }
+
+    // Currently only concerned with instructions that follow movprfx
+    if (firstId->idIns() != INS_sve_movprfx)
+    {
+        return;
+    }
+
+    // Quoted sections are taken fron the Arm Arm.
+
+    // "It is required that the prefixed instruction at PC+4 must be an SVE destructive binary or ternary
+    // instruction encoding, or a unary operation with merging predication, but excluding other MOVPRFX instructions."
+    // "The prefixed instruction must not use the destination register in any other operand position, even if
+    // they have different names but refer to the same architectural register state."
+    switch (secondId->idInsFmt())
+    {
+        case IF_SVE_AM_2A:   // <Zdn>.<T>, <Pg>/M, <Zdn>.<T>, #<const>
+        case IF_SVE_BN_1A:   // <Zdn>.D{, <pattern>{, MUL #<imm>}}
+        case IF_SVE_BP_1A:   // <Zdn>.D{, <pattern>{, MUL #<imm>}}
+        case IF_SVE_BS_1A:   // <Zdn>.<T>, <Zdn>.<T>, #<const>
+        case IF_SVE_BU_2A:   // <Zd>.<T>, <Pg>/M, #<const>
+        case IF_SVE_BV_2A_A: // <Zd>.<T>, <Pg>/M, #<imm>{, <shift>}
+        case IF_SVE_BV_2A_J: // <Zd>.<T>, <Pg>/M, #<imm>{, <shift>}
+        case IF_SVE_BV_2B:   // <Zd>.<T>, <Pg>/M, #0.0
+        case IF_SVE_CD_2A:   // <Zdn>.<T>, <R><m>
+        case IF_SVE_CQ_3A:   // <Zd>.<T>, <Pg>/M, <R><n|SP>
+        case IF_SVE_DN_2A:   // <Zdn>.<T>, <Pm>.<T>
+        case IF_SVE_DP_2A:   // <Zdn>.<T>, <Pm>.<T>
+        case IF_SVE_EC_1A:   // <Zdn>.<T>, <Zdn>.<T>, #<imm>{, <shift>}
+        case IF_SVE_ED_1A:   // <Zdn>.<T>, <Zdn>.<T>, #<imm>
+        case IF_SVE_EE_1A:   // <Zdn>.<T>, <Zdn>.<T>, #<imm>
+        case IF_SVE_HM_2A:   // <Zdn>.<T>, <Pg>/M, <Zdn>.<T>, <const>
+            break;
+
+        case IF_SVE_CC_2A: // <Zdn>.<T>, <V><m>
+        case IF_SVE_FU_2A: // <Zda>.<T>, <Zn>.<T>, #<const>
+            assert(secondId->idReg1() != secondId->idReg2());
```

Comment:
> i was hoping to see lot more validation for registers of firstId

That should be all at the bottom of the function. I'll double check that I have everything, because the text you have above is a much more verbose version than what I was using.

Reply:
looks like you have most of the things covered. We need to add validation for:

- MOVPRFX cannot be the last instruction in a sequence
- The predicated instruction at PC+4 must use the merging predicate. (not sure if this is taken care by the format checks)
- MOVPRFX destination register MUST be used as the destination register in the instruction at PC+4, "**and is not allowed to be used in any other position other than destructive input**"

Reply:
> * MOVPRFX cannot be the last instruction in a sequence

Done.

> * The predicated instruction at PC+4 must use the merging predicate. (not sure if this is taken care by the format checks)

Looking at all the formats, they are all <Pg>/M or unpredicated. I've added some extra checking to make sure they correspond to the correct movprfx instruction.

> * MOVPRFX destination register MUST be used as the destination register in the instruction at PC+4, "**and is not allowed to be used in any other position other than destructive input**"

That's done in the switch. Eg:
`            assert(secondId->idReg1() != secondId->idReg3());`
I had to do it there because you need the format to know what types the registers are.